onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc

```diff
@@ -89,39 +118,64 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputP
+      LOGS(logger, VERBOSE) << "Reshape does not support new shape with 0. "
+                            << "New shape: " << Shape2String(new_shape);
+      return false;
+    }
+  }
+
   std::vector<int64_t> input_shape;
-  if (!GetStaticShape(*input_defs[0], input_shape, logger))
-    return false;
 
-  if (input_shape.empty()) {
-    LOGS(logger, VERBOSE) << "Reshape does not support empty input shape";
+  // first input must be fixed rank OR (first input has variadic rank AND shape only contains positive integers)
+  // as per docs, 0 is considered an illegal shape element if the input is variadic
+  if (!GetShape(*input_defs[0], input_shape, logger)) {
+    if (!AllPositiveShape(new_shape)) {
+      LOGS(logger, VERBOSE) << "Unable to get shape of input -- input must have fixed rank for reshape, unless the new shape contains all positive integers. "
+                               "Input shape: "
+                            << Shape2String(input_shape) << ", new shape: " << Shape2String(new_shape);
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  if (!input_params.create_mlprogram && !IsStaticShape(input_shape)) {
+    LOGS(logger, VERBOSE) << "Input must have static shape for NeuralNetwork.";
     return false;
   }
 
-  // CoreML reshape doesn't support new shape with more than 5 dimensions.
+  if (std::find(input_shape.begin(), input_shape.end(), 0) != input_shape.end()) {
+    LOGS(logger, VERBOSE) << "Input shape must not have any 0's in it.";
+  }
+
   if (new_shape.size() > 5) {
-    LOGS(logger, VERBOSE) << "Reshape does not support new shape with rank greater than 5. Input shape: "
+    LOGS(logger, VERBOSE) << "Reshape does not support new shape with more than 5 dimensions. "
+                             "Input shape: "
                           << Shape2String(input_shape) << ", new shape: " << Shape2String(new_shape);
     return false;
   }
 
-  // CoreML reshape does not support 0 as dimension
-  NodeAttrHelper helper(node);
-  const bool allow_zero = helper.Get("allowzero", 0) == 1;
-  if (allow_zero) {
-    if (std::find(new_shape.begin(), new_shape.end(), int64_t{0}) != new_shape.end()) {
-      LOGS(logger, VERBOSE) << "Reshape does not support new shape with 0 as dimension when allowzero is enabled. "
-                               "Input shape: "
-                            << Shape2String(input_shape) << ", new shape: " << Shape2String(new_shape);
-      return false;
-    }
+  if (!CheckNegativeOneCollision(input_shape, new_shape, logger)) {
+    return false;
   }
 
+  // Max of one -1 allowed
+  if (input_params.create_mlprogram && std::count(new_shape.begin(), new_shape.end(), -1) > 1) {
+    LOGS(logger, VERBOSE) << "Reshape does not support -1 in new shape unless it appears in the input shape. "
+                             "Input shape: "
+                          << Shape2String(input_shape) << ", new shape: " << Shape2String(new_shape);
+    return false;
```

Comment:
nit: the onnx type/shape inferencing should have enforced this as we know the `shape` input is a const initializer.

https://github.com/onnx/onnx/blob/a570866ba4641506d5530b9b559e169ded3edfe1/onnx/defs/tensor/defs.cc#L289-L291